velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -275,6 +317,15 @@ class AggregateWindowFunction : public exec::WindowFunction {
   char* rawSingleGroupRow_;
   vector_size_t singleGroupRowSize_;
 
+  // This offset tracks how far along the partition rows have been output.
+  // This can be used to optimize reading offset column values corresponding
+  // to the present row set in getOutput.
+  vector_size_t partitionOffset_;
+  // Tracks if a frame bound of 'between unbounded preceding and k preceding' is
+  // seen in the aggregate function call. This is used to adjust the number of
+  // rows read to compute the aggregate.
+  bool hasEmptyPrecedingFrame_;
```

Comment:
Why is this a member variable ?

Reply:
When an empty preceding frame is seen in the first call to apply, the startRow and numFrameRows should be changed accordingly in all subsequent calls for correctness of the aggregate calculation.